File: C++/Tokenmap/Tokenmap.cpp
#include "Tokennode.h"
#include "Tokenlinkedlist.h"

#include<iostream>
#include<malloc.h>
#include<string>
#include<vector>

using std::cout;
using std::cin;
using std::end;
using std::string;
using std::vector;


class Tokenmap{

            public:
                    int size;
                    int num_elements;
                    vector<Tokenlinkedlist> slots;

                    Tokenmap(int s):slots(s),size(s),num_elements(0){}

                    void print_tokenmap(Tokenmap& tm){
                        if(tm.num_elements == 0)
                        cout << "The token map is empty \n";
                        else{
                            for(int i=0;i<tm.size;i++){
                                if(tm.slots[i].head == nullptr)
                                cout << "This is slot " << i << " and it is empty \n";
                                else{
                                    cout << "This is slot "+ i << " and it contains the following \n";
                                    Tokennode *aux = tm.slots[i].head;
                                    while(aux != nullptr){
                                        cout << "Tokend: " << aux->token <<" id: "<< aux->id << " || ";
                                        aux = aux->next;
                                    }
                                    cout << "\n";   
                                }
                            }
                        }
                    }


                    int hashcode(Tokenmap& tm,string t){
                        int c = 0;

                        int temp = 0;
                        for(int i=0;i<t.length();i++)
                        temp = temp + (t[i] -'0');
                        
                        c = ((temp << 2)^temp)%tm.size;

                        return c;
                    }

                    void resize(Tokenmap& tm){
                        
                    }

                    void insert_tokenid(Tokenmap& tm,string t,int num){
                        if(tm.num_elements >= tm.size)
                        tm.resize(tm);

                        int c = tm.hashcode(tm,t);
                        Tokennode *n = new Tokennode(t,num);
                        
                        Tokennode *aux = tm.slots[c].head;
                        if(aux == nullptr)
                        tm.slots[c].head = n;
                        else{
                            while(aux->next != nullptr)
                            aux = aux->next;
                            
                            aux->next = n;
                            tm.num_elements++;
                        }
                    }

                    void remove_idtoken(Tokenmap& tm,string t){
                            int c = tm.hashcode(tm,t);
                            Tokennode *aux = tm.slots[c].head;
                            Tokennode *temp = nullptr;

                            bool b = false;
                            while(aux!= nullptr && !b){
                                if(aux->token == t)
                                b = true;
                                else{
                                    temp = aux;
                                    aux = aux->next;
                                }
                            }

                            if(b){
                                if(temp == nullptr)
                                tm.slots[c].head = tm.slots[c].head->next;
                                else temp->next = aux->next;

                                tm.num_elements--;
                            }else cout << "This token does not exist"; 
                    }

                    Tokennode retrieve_idtoken(Tokenmap& tm,string t){
                            int c = tm.hashcode(tm,t);
                            
                            Tokennode *aux = tm.slots[c].head;
                            bool b = false;
                            while( aux != nullptr && !b){
                                if(aux->token == t)
                                b = true;
                                else aux = aux->next;
                            }

                            Tokennode r = Tokennode("",-1);
                            if(b)
                            r = Tokennode(aux->token,aux->id);

                        return r;
                    }   
};

int main(){

    cout << "Test \n";


    exit(1);
}

File: Python/Tokenmap/Tokenmap.py
import os
import sys
import numpy as np

from .Tokennode import Tokennode
from .Tokenlinkedlist import Tokenlinkedlist

class Tokenmap:
    def __init__(self,s):
        self.size = s
        self.num_tokens = 0
        self.slots = s*[None]

        i = 0
        while i < self.size:
            self.slots[i] = Tokenlinkedlist()
            i+=1
            
    def hashcode(self,t):
        i = 0
        temp = 0
        code = 0
        while i < len(t):
            temp = temp + ord(t[i])
            i+=1

        code = ((temp << 2)^temp)%self.size

        return code
    
    def printTokenmap(self):
        if self.num_tokens == 0:
            print("The token map is empty \n")
        else:
            i = 0
            while i < self.size:
                if self.slots[i].head == None:
                    print("Slot ",i," is empty")
                else:
                    print("Slot ",i," contains the following values:")
                    aux = self.slots[i].head
                    while aux != None:
                        print("Token: ",aux.token," || ID: ",aux.id," || ",end="")
                        aux = aux.next
                    
                    print()
                
                i+=1

    def resize(self):
        tempmap = Tokenmap(2*self.size)
        
        i = 0
        while i < self.size:
            if self.slots[i].head != None:
                aux = self.slots[i].head
                while aux != None:
                    tempmap.add_token(aux.token,aux.id)
                    aux = aux.next 

            i+=1
        
        self.size = 2*self.size
        self.slots = tempmap.slots

    def add_token(self,t,id):
        r_t = self.retrieve_token(t)

        if r_t == None:
            if self.num_tokens >= self.size or ((self.num_tokens/self.size)*100) >= 70:
                self.resize()
        
            temp = Tokennode(t,id)
            c = self.hashcode(t)
            aux = self.slots[c].head
            if aux == None:
                self.slots[c].head = temp
            else:
                while aux.next != None:
                    aux = aux.next
                
                aux.next = temp
            
            self.num_tokens += 1
    
    def remove_token(self,t):

        c = self.hashcode(t)
        aux = self.slots[c].head

        b = False
        temp = None
        while aux != None and not b:
            if aux.token == t:
                b = True
            else:
                temp = aux
                aux = aux.next

        if b:
            if temp == None:
                self.slots[c].head = self.slots[c].head.next
            else:
                temp.next = aux.next

            self.num_tokens-=1

    
    def retrieve_token(self,t):
        r = None
 
        c = self.hashcode(t)
        aux = self.slots[c].head
        b = False

        while aux != None and not b:
            if aux.token == t:
                b = True
            else:
                aux = aux.next
        
        if b:
            r = aux

        return r

    def update_token(self,t,v):
        c = self.hashcode(t)

        aux = self.slots[c].head
        b = False
        while aux != None and not b:
            if aux.token == t:
                b = True
            else:
                aux = aux.next

        if b:
            aux.id = v
        


"""
print("Give me the number of tokens you would to insert: \n")
s = int(input())
tm = Tokenmap(s)

print("Give me your tokens and IDs: \n")
i = 0
while i < s:
    t = input()
    num_id = int(input())
    tm.add_token(t,num_id)
    i+=1

print()
tm.printTokenmap()
print()

while True:
    k = tm.num_tokens
    print()
    print("Give me the entry that you would like to update and the new value:")
    e = input()
    x = int(input())
    print()
    tm.update_token(e,x)
    tm.printTokenmap()
    print()
"""